{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Data retrieval from GEO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this exercise, we are downloading data from the NCBI GEO database via programmatic access. This exercise is based on the example from https://geoparse.readthedocs.io/en/latest/usage.html#examples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GEOparse is a Python library to access Gene Expression Omnibus Database (GEO). GEOparse.get_GEO() will check the GEO database for a specified accession ID and download it to specified directory. The result will be loaded into GEOparse.GSE file. See the documentation in https://geoparse.readthedocs.io/en/latest/introduction.html#features.\n",
    "\n",
    "\n",
    "We will get familiar with exploring unfamiliar data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation of libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first step is to import the required Python libraries. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip is the package installer for Python, see https://pypi.org/project/pip/ for details\n",
    "#\n",
    "#import sys\n",
    "#!{sys.executable} -m pip install GEOparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "# To read, write and process tabular data:\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's download an example data set from the study \"Kidney Transplant Rejection and Tissue Injury by Gene Profiling of Biopsies and Peripheral Blood Lymphocytes\" by Flechner et al, 2007 (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2041877/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\Peder\\\\Documents\\\\repos\\\\CBM-101\\\\CBM101\\\\C_Data_resources'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check your current working folder if necessary:\n",
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "28-Oct-2020 13:28:20 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "28-Oct-2020 13:28:20 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE1nnn/GSE1563/soft/GSE1563_family.soft.gz to ./GSE1563_family.soft.gz\n",
      "100%|█████████████████████████████████████████████████████████████████████████████| 9.62M/9.62M [00:05<00:00, 1.74MB/s]\n",
      "28-Oct-2020 13:28:28 DEBUG downloader - Size validation passed\n",
      "28-Oct-2020 13:28:28 DEBUG downloader - Moving C:\\Users\\Peder\\AppData\\Local\\Temp\\tmpb1p_l7ge to C:\\Users\\Peder\\Documents\\repos\\CBM-101\\CBM101\\C_Data_resources\\GSE1563_family.soft.gz\n",
      "28-Oct-2020 13:28:28 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE1nnn/GSE1563/soft/GSE1563_family.soft.gz\n",
      "28-Oct-2020 13:28:28 INFO GEOparse - Parsing ./GSE1563_family.soft.gz: \n",
      "28-Oct-2020 13:28:28 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "28-Oct-2020 13:28:28 DEBUG GEOparse - SERIES: GSE1563\n",
      "28-Oct-2020 13:28:28 DEBUG GEOparse - PLATFORM: GPL8300\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26805\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26806\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26807\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26808\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26809\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26810\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26811\n",
      "28-Oct-2020 13:28:30 DEBUG GEOparse - SAMPLE: GSM26812\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26813\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26814\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26815\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26816\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26817\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26818\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26819\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26820\n",
      "28-Oct-2020 13:28:31 DEBUG GEOparse - SAMPLE: GSM26821\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26822\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26823\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26824\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26825\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26826\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26827\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26828\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26829\n",
      "28-Oct-2020 13:28:32 DEBUG GEOparse - SAMPLE: GSM26830\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26831\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26832\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26833\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26834\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26835\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26836\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26837\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26838\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26839\n",
      "28-Oct-2020 13:28:33 DEBUG GEOparse - SAMPLE: GSM26840\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26841\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26842\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26843\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26844\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26845\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26846\n",
      "28-Oct-2020 13:28:34 DEBUG GEOparse - SAMPLE: GSM26847\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26848\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26849\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26850\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26851\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26852\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26853\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26854\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26855\n",
      "28-Oct-2020 13:28:35 DEBUG GEOparse - SAMPLE: GSM26856\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26857\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26858\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26859\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26860\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26861\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26862\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26863\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26864\n",
      "28-Oct-2020 13:28:36 DEBUG GEOparse - SAMPLE: GSM26865\n",
      "28-Oct-2020 13:28:37 DEBUG GEOparse - SAMPLE: GSM26866\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM example:\n",
      "Name:  GSM26805\n",
      "Metadata:\n",
      " - title : C1PBL\n",
      " - geo_accession : GSM26805\n",
      " - status : Public on Jul 14 2004\n",
      " - submission_date : Jul 14 2004\n",
      " - last_update_date : Mar 16 2009\n",
      " - type : RNA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : PBL\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - molecule_ch1 : total RNA\n",
      " - description : Clinical status: control healthy blood donor, Age: unknown, Sex: unknown, Immunosupression: none, Histopathology: none, Donor type: NA, Scr (mg/dL): unknown, Days post transplant: NA, Abbreviations used in sample description: Abreviations used to describe patient samples include the following: BX - Biopsy; PBL- Peripheral Blood Lymphocytes; CsA -Cyclosporine; MMF - Mycophenolate Mofetil; P - Prednisone; FK - Tacrolimus;  SRL - Sirolimus; CAD -Cadaveric;  LD - Live Donor; Scr - Serum Creatinine; ATN - Acute Tubular Necrosis CNI - Calcineurin Inhibitor; FSGS - Focal Segmental Glomerulosclerosis, Keywords = DNA microarrays, gene expression, kidney, rejection, transplant\n",
      " - platform_id : GPL8300\n",
      " - contact_name : Steven,,Head\n",
      " - contact_email : shead@scripps.edu\n",
      " - contact_phone : 858 784-2263\n",
      " - contact_institute : The Scripps Research Institute\n",
      " - contact_address : 10550 N Torrey Pines Rd\n",
      " - contact_city : La Jolla\n",
      " - contact_state : CA\n",
      " - contact_zip/postal_code : 92037\n",
      " - contact_country : USA\n",
      " - supplementary_file : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM26nnn/GSM26805/suppl/GSM26805.CEL.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM26nnn/GSM26805/suppl/GSM26805.EXP.gz\n",
      " - series_id : GSE1563\n",
      " - data_row_count : 12625\n",
      "Table data:\n",
      "            ID_REF  VALUE ABS_CALL  DETECTION P-VALUE\n",
      "0   AFFX-MurIL2_at  211.5        A            0.50000\n",
      "1  AFFX-MurIL10_at  190.2        A            0.23906\n",
      "2   AFFX-MurIL4_at   96.5        A            0.68628\n",
      "3   AFFX-MurFAS_at   80.8        A            0.55935\n",
      "4   AFFX-BioB-5_at  847.3        P            0.00202\n"
     ]
    }
   ],
   "source": [
    "# download the data set using GEOparse(the data is available in GEO database with the accession ID GSE1563)\n",
    "\n",
    "kidney_data = GEOparse.get_GEO(geo=\"GSE1563\", destdir=\"./\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 1. Inspect your downloaded data. a) what data type is it?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<SERIES: GSE1563 - 62 SAMPLES, 1 d(s)>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "GEOparse.GEOTypes.GSE"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %load solutions/ex1_1a.py\n",
    "print(kidney_data)\n",
    "type(kidney_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b) what does it contain? Try to play around to access these different contents.\n",
    "Hint: use `dir` or write `kidney_data.` and press Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE1563 \n",
      "\n",
      "Expression profiling by array \n",
      "\n",
      "!Series_title = Kidney Transplant Rejection and Tissue Injury by Gene Profiling of Biopsies and Peripheral Blood Lymphocytes\n",
      "!Series_geo_accession = GSE1563\n",
      "!Series_status = Public on Jul 14 2004\n",
      "!Series_submission_date = Jul 14 2004\n",
      "!Series_last_update_date = Dec 13 2018\n",
      "!Series_pubmed_id = 15307835\n",
      "!Series_summary = We used DNA microarrays (HG-U95Av2 GeneChips) to determine gene expression profiles for kidney biopsies and peripheral blood lymphocytes (PBLs) in transplant patients. Sample classes include kidney biopsies and PBLs from patients with 1) healthy normal donor kidneys, 2) well-functioning transplants with no clinical evidence of rejection, 3) kidneys undergoing acute rejection, and 4) transplants with renal dysfunction without rejection. Nomenclature for samples is as follows: 1) all sample names include either BX or PBL to indicate that they were derived from biopsies or PBLs respectively, 2) C indicates samples from healthy normal donors, 3) TX indicates samples from patients with well-functioning transplants with no clinical evidence of rejection, 3) AR indicates samples from transplant patients with kidneys undergoing acute rejection, 4) NR indicates samples from transplant patients with renal dysfunction without rejection.\n",
      "!Series_summary = Abbreviations used to describe patient samples include the following: BX - Biopsy; PBL- Peripheral Blood Lymphocytes; CsA -Cyclosporine; MMF - Mycophenolate Mofetil; P - Prednisone; FK - Tacrolimus;  SRL - Sirolimus; CAD -Cadaveric;  LD - Live Donor; Scr - Serum Creatinine; ATN - Acute Tubular Necrosis CNI - Calcineurin Inhibitor; FSGS - Focal Segmental Glomerulosclerosis\n",
      "!Series_summary = several array data sets did not pass quality control and were not analyzed. These include AR1PBL, NR4BX, and NR6PBL\n",
      "!Series_summary = Keywords = DNA microarrays, gene expression, kidney, rejection, transplant\n",
      "!Series_summary = Keywords: other\n",
      "!Series_type = Expression profiling by array\n",
      "!Series_contributor = Stuart,M,Flechner\n",
      "!Series_contributor = Sunil,M,Kurian\n",
      "!Series_contributor = Steven,R,Head\n",
      "!Series_contributor = Starlette,M,Sharp\n",
      "!Series_contributor = Thomas,C,Whisenant\n",
      "!Series_contributor = Jie,,Zhang\n",
      "!Series_contributor = Jeffrey,D,Chismar\n",
      "!Series_contributor = Steve,,Horvath\n",
      "!Series_contributor = Tony,,Mondala\n",
      "!Series_contributor = Timothy,,Gilmartin\n",
      "!Series_contributor = Daniel,J,Cook\n",
      "!Series_contributor = Steven,A,Kay\n",
      "!Series_contributor = John,R,Walker\n",
      "!Series_contributor = Daniel,R,Salomon\n",
      "!Series_sample_id = GSM26805\n",
      "!Series_sample_id = GSM26806\n",
      "!Series_sample_id = GSM26807\n",
      "!Series_sample_id = GSM26808\n",
      "!Series_sample_id = GSM26809\n",
      "!Series_sample_id = GSM26810\n",
      "!Series_sample_id = GSM26811\n",
      "!Series_sample_id = GSM26812\n",
      "!Series_sample_id = GSM26813\n",
      "!Series_sample_id = GSM26814\n",
      "!Series_sample_id = GSM26815\n",
      "!Series_sample_id = GSM26816\n",
      "!Series_sample_id = GSM26817\n",
      "!Series_sample_id = GSM26818\n",
      "!Series_sample_id = GSM26819\n",
      "!Series_sample_id = GSM26820\n",
      "!Series_sample_id = GSM26821\n",
      "!Series_sample_id = GSM26822\n",
      "!Series_sample_id = GSM26823\n",
      "!Series_sample_id = GSM26824\n",
      "!Series_sample_id = GSM26825\n",
      "!Series_sample_id = GSM26826\n",
      "!Series_sample_id = GSM26827\n",
      "!Series_sample_id = GSM26828\n",
      "!Series_sample_id = GSM26829\n",
      "!Series_sample_id = GSM26830\n",
      "!Series_sample_id = GSM26831\n",
      "!Series_sample_id = GSM26832\n",
      "!Series_sample_id = GSM26833\n",
      "!Series_sample_id = GSM26834\n",
      "!Series_sample_id = GSM26835\n",
      "!Series_sample_id = GSM26836\n",
      "!Series_sample_id = GSM26837\n",
      "!Series_sample_id = GSM26838\n",
      "!Series_sample_id = GSM26839\n",
      "!Series_sample_id = GSM26840\n",
      "!Series_sample_id = GSM26841\n",
      "!Series_sample_id = GSM26842\n",
      "!Series_sample_id = GSM26843\n",
      "!Series_sample_id = GSM26844\n",
      "!Series_sample_id = GSM26845\n",
      "!Series_sample_id = GSM26846\n",
      "!Series_sample_id = GSM26847\n",
      "!Series_sample_id = GSM26848\n",
      "!Series_sample_id = GSM26849\n",
      "!Series_sample_id = GSM26850\n",
      "!Series_sample_id = GSM26851\n",
      "!Series_sample_id = GSM26852\n",
      "!Series_sample_id = GSM26853\n",
      "!Series_sample_id = GSM26854\n",
      "!Series_sample_id = GSM26855\n",
      "!Series_sample_id = GSM26856\n",
      "!Series_sample_id = GSM26857\n",
      "!Series_sample_id = GSM26858\n",
      "!Series_sample_id = GSM26859\n",
      "!Series_sample_id = GSM26860\n",
      "!Series_sample_id = GSM26861\n",
      "!Series_sample_id = GSM26862\n",
      "!Series_sample_id = GSM26863\n",
      "!Series_sample_id = GSM26864\n",
      "!Series_sample_id = GSM26865\n",
      "!Series_sample_id = GSM26866\n",
      "!Series_contact_name = Steven,,Head\n",
      "!Series_contact_email = shead@scripps.edu\n",
      "!Series_contact_phone = 858 784-2263\n",
      "!Series_contact_institute = The Scripps Research Institute\n",
      "!Series_contact_address = 10550 N Torrey Pines Rd\n",
      "!Series_contact_city = La Jolla\n",
      "!Series_contact_state = CA\n",
      "!Series_contact_zip/postal_code = 92037\n",
      "!Series_contact_country = USA\n",
      "!Series_supplementary_file = ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE1nnn/GSE1563/suppl/GSE1563_RAW.tar\n",
      "!Series_platform_id = GPL8300\n",
      "!Series_platform_taxid = 9606\n",
      "!Series_sample_taxid = 9606\n",
      "!Series_relation = BioProject: https://www.ncbi.nlm.nih.gov/bioproject/PRJNA90879\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# %load solutions/ex1_1b.py\n",
    "dir(kidney_data)\n",
    "\n",
    "# for example: \n",
    "print(kidney_data.name, '\\n')\n",
    "print(kidney_data.get_type(), '\\n')\n",
    "print(kidney_data.show_metadata())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### c) look into the GSMs of `kidney_data`. \n",
    "Hint: you can also use the Tab trick multiple times to go deeper e.g. `kidney_data.gsms.` and press Tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GSM26805': <SAMPLE: GSM26805>,\n",
       " 'GSM26806': <SAMPLE: GSM26806>,\n",
       " 'GSM26807': <SAMPLE: GSM26807>,\n",
       " 'GSM26808': <SAMPLE: GSM26808>,\n",
       " 'GSM26809': <SAMPLE: GSM26809>,\n",
       " 'GSM26810': <SAMPLE: GSM26810>,\n",
       " 'GSM26811': <SAMPLE: GSM26811>,\n",
       " 'GSM26812': <SAMPLE: GSM26812>,\n",
       " 'GSM26813': <SAMPLE: GSM26813>,\n",
       " 'GSM26814': <SAMPLE: GSM26814>,\n",
       " 'GSM26815': <SAMPLE: GSM26815>,\n",
       " 'GSM26816': <SAMPLE: GSM26816>,\n",
       " 'GSM26817': <SAMPLE: GSM26817>,\n",
       " 'GSM26818': <SAMPLE: GSM26818>,\n",
       " 'GSM26819': <SAMPLE: GSM26819>,\n",
       " 'GSM26820': <SAMPLE: GSM26820>,\n",
       " 'GSM26821': <SAMPLE: GSM26821>,\n",
       " 'GSM26822': <SAMPLE: GSM26822>,\n",
       " 'GSM26823': <SAMPLE: GSM26823>,\n",
       " 'GSM26824': <SAMPLE: GSM26824>,\n",
       " 'GSM26825': <SAMPLE: GSM26825>,\n",
       " 'GSM26826': <SAMPLE: GSM26826>,\n",
       " 'GSM26827': <SAMPLE: GSM26827>,\n",
       " 'GSM26828': <SAMPLE: GSM26828>,\n",
       " 'GSM26829': <SAMPLE: GSM26829>,\n",
       " 'GSM26830': <SAMPLE: GSM26830>,\n",
       " 'GSM26831': <SAMPLE: GSM26831>,\n",
       " 'GSM26832': <SAMPLE: GSM26832>,\n",
       " 'GSM26833': <SAMPLE: GSM26833>,\n",
       " 'GSM26834': <SAMPLE: GSM26834>,\n",
       " 'GSM26835': <SAMPLE: GSM26835>,\n",
       " 'GSM26836': <SAMPLE: GSM26836>,\n",
       " 'GSM26837': <SAMPLE: GSM26837>,\n",
       " 'GSM26838': <SAMPLE: GSM26838>,\n",
       " 'GSM26839': <SAMPLE: GSM26839>,\n",
       " 'GSM26840': <SAMPLE: GSM26840>,\n",
       " 'GSM26841': <SAMPLE: GSM26841>,\n",
       " 'GSM26842': <SAMPLE: GSM26842>,\n",
       " 'GSM26843': <SAMPLE: GSM26843>,\n",
       " 'GSM26844': <SAMPLE: GSM26844>,\n",
       " 'GSM26845': <SAMPLE: GSM26845>,\n",
       " 'GSM26846': <SAMPLE: GSM26846>,\n",
       " 'GSM26847': <SAMPLE: GSM26847>,\n",
       " 'GSM26848': <SAMPLE: GSM26848>,\n",
       " 'GSM26849': <SAMPLE: GSM26849>,\n",
       " 'GSM26850': <SAMPLE: GSM26850>,\n",
       " 'GSM26851': <SAMPLE: GSM26851>,\n",
       " 'GSM26852': <SAMPLE: GSM26852>,\n",
       " 'GSM26853': <SAMPLE: GSM26853>,\n",
       " 'GSM26854': <SAMPLE: GSM26854>,\n",
       " 'GSM26855': <SAMPLE: GSM26855>,\n",
       " 'GSM26856': <SAMPLE: GSM26856>,\n",
       " 'GSM26857': <SAMPLE: GSM26857>,\n",
       " 'GSM26858': <SAMPLE: GSM26858>,\n",
       " 'GSM26859': <SAMPLE: GSM26859>,\n",
       " 'GSM26860': <SAMPLE: GSM26860>,\n",
       " 'GSM26861': <SAMPLE: GSM26861>,\n",
       " 'GSM26862': <SAMPLE: GSM26862>,\n",
       " 'GSM26863': <SAMPLE: GSM26863>,\n",
       " 'GSM26864': <SAMPLE: GSM26864>,\n",
       " 'GSM26865': <SAMPLE: GSM26865>,\n",
       " 'GSM26866': <SAMPLE: GSM26866>}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kidney_data.gsms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'GEOparse.GEOTypes.GSM'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'title': ['C1PBL'],\n",
       " 'geo_accession': ['GSM26805'],\n",
       " 'status': ['Public on Jul 14 2004'],\n",
       " 'submission_date': ['Jul 14 2004'],\n",
       " 'last_update_date': ['Mar 16 2009'],\n",
       " 'type': ['RNA'],\n",
       " 'channel_count': ['1'],\n",
       " 'source_name_ch1': ['PBL'],\n",
       " 'organism_ch1': ['Homo sapiens'],\n",
       " 'taxid_ch1': ['9606'],\n",
       " 'molecule_ch1': ['total RNA'],\n",
       " 'description': ['Clinical status: control healthy blood donor',\n",
       "  'Age: unknown',\n",
       "  'Sex: unknown',\n",
       "  'Immunosupression: none',\n",
       "  'Histopathology: none',\n",
       "  'Donor type: NA',\n",
       "  'Scr (mg/dL): unknown',\n",
       "  'Days post transplant: NA',\n",
       "  'Abbreviations used in sample description: Abreviations used to describe patient samples include the following: BX - Biopsy; PBL- Peripheral Blood Lymphocytes; CsA -Cyclosporine; MMF - Mycophenolate Mofetil; P - Prednisone; FK - Tacrolimus;  SRL - Sirolimus; CAD -Cadaveric;  LD - Live Donor; Scr - Serum Creatinine; ATN - Acute Tubular Necrosis CNI - Calcineurin Inhibitor; FSGS - Focal Segmental Glomerulosclerosis',\n",
       "  'Keywords = DNA microarrays, gene expression, kidney, rejection, transplant'],\n",
       " 'platform_id': ['GPL8300'],\n",
       " 'contact_name': ['Steven,,Head'],\n",
       " 'contact_email': ['shead@scripps.edu'],\n",
       " 'contact_phone': ['858 784-2263'],\n",
       " 'contact_institute': ['The Scripps Research Institute'],\n",
       " 'contact_address': ['10550 N Torrey Pines Rd'],\n",
       " 'contact_city': ['La Jolla'],\n",
       " 'contact_state': ['CA'],\n",
       " 'contact_zip/postal_code': ['92037'],\n",
       " 'contact_country': ['USA'],\n",
       " 'supplementary_file': ['ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM26nnn/GSM26805/suppl/GSM26805.CEL.gz',\n",
       "  'ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM26nnn/GSM26805/suppl/GSM26805.EXP.gz'],\n",
       " 'series_id': ['GSE1563'],\n",
       " 'data_row_count': ['12625']}"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %load solutions/ex1_1c.py\n",
    "\n",
    "# by usingt he above trick we see that kidney_data.gsms is a dictionary (you can validate this)\n",
    "\n",
    "type(kidney_data.gsms)\n",
    "\n",
    "# a dict is a container of key,value pairs. The values in this case are of the class GEOparse.GEOTypes.GSM \n",
    "# e.g.\n",
    "print(type(list(kidney_data.gsms.values())[0]))\n",
    "\n",
    "## to get the available methods:\n",
    "fst = list(kidney_data.gsms.values())[0]\n",
    "dir(fst)\n",
    "\n",
    "# for instance look at metadata:\n",
    "\n",
    "fst.metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Printing a summary\n",
    "We could then do something like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM example:\n",
      "-------------\n",
      "Name:  GSM26805\n",
      "Metadata:\n",
      " - title : C1PBL\n",
      " - geo_accession : GSM26805\n",
      " - status : Public on Jul 14 2004\n",
      " - submission_date : Jul 14 2004\n",
      " - last_update_date : Mar 16 2009\n",
      " - type : RNA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : PBL\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - molecule_ch1 : total RNA\n",
      " - description : Clinical status: control healthy blood donor, Age: unknown, Sex: unknown, Immunosupression: none, Histopathology: none, Donor type: NA, Scr (mg/dL): unknown, Days post transplant: NA, Abbreviations used in sample description: Abreviations used to describe patient samples include the following: BX - Biopsy; PBL- Peripheral Blood Lymphocytes; CsA -Cyclosporine; MMF - Mycophenolate Mofetil; P - Prednisone; FK - Tacrolimus;  SRL - Sirolimus; CAD -Cadaveric;  LD - Live Donor; Scr - Serum Creatinine; ATN - Acute Tubular Necrosis CNI - Calcineurin Inhibitor; FSGS - Focal Segmental Glomerulosclerosis, Keywords = DNA microarrays, gene expression, kidney, rejection, transplant\n",
      " - platform_id : GPL8300\n",
      " - contact_name : Steven,,Head\n",
      " - contact_email : shead@scripps.edu\n",
      " - contact_phone : 858 784-2263\n",
      " - contact_institute : The Scripps Research Institute\n",
      " - contact_address : 10550 N Torrey Pines Rd\n",
      " - contact_city : La Jolla\n",
      " - contact_state : CA\n",
      " - contact_zip/postal_code : 92037\n",
      " - contact_country : USA\n",
      " - supplementary_file : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM26nnn/GSM26805/suppl/GSM26805.CEL.gz, ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM26nnn/GSM26805/suppl/GSM26805.EXP.gz\n",
      " - series_id : GSE1563\n",
      " - data_row_count : 12625\n",
      "Table data:\n",
      "\n",
      "            ID_REF  VALUE ABS_CALL  DETECTION P-VALUE\n",
      "0   AFFX-MurIL2_at  211.5        A            0.50000\n",
      "1  AFFX-MurIL10_at  190.2        A            0.23906\n",
      "2   AFFX-MurIL4_at   96.5        A            0.68628\n",
      "3   AFFX-MurFAS_at   80.8        A            0.55935\n",
      "4   AFFX-BioB-5_at  847.3        P            0.00202\n"
     ]
    }
   ],
   "source": [
    "# A GSM (or a Sample) contains information the conditions and preparation of the sample\n",
    "\n",
    "print(\"GSM example:\\n-------------\")\n",
    "for gsm_name, gsm in kidney_data.gsms.items():\n",
    "    print(\"Name: \", gsm_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gsm.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print (\"Table data:\",)\n",
    "    print()\n",
    "    print (gsm.table.head())\n",
    "    break # so we stop after the first\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GPL example:\n",
      "-------------\n",
      "Name:  GPL8300\n",
      "Metadata:\n",
      " - title : [HG_U95Av2] Affymetrix Human Genome U95 Version 2 Array\n",
      " - geo_accession : GPL8300\n",
      " - status : Public on Mar 16 2009\n",
      " - submission_date : Mar 13 2009\n",
      " - last_update_date : Dec 13 2018\n",
      " - technology : in situ oligonucleotide\n",
      " - distribution : commercial\n",
      " - organism : Homo sapiens\n",
      " - taxid : 9606\n",
      " - manufacturer : Affymetrix\n",
      " - manufacture_protocol : see manufacturer's web site, , Based on this UniGene build and associated annotations, the HG-U95Av2 array represents approximately 10,000 full-length genes., \n",
      " - description : Affymetrix submissions are typically submitted to GEO using the GEOarchive method described at http://www.ncbi.nlm.nih.gov/projects/geo/info/geo_affy.html, , June 03, 2009: annotation table updated with netaffx build 28, June 08, 2012: annotation table updated with netaffx build 32, June 27, 2016: annotation table updated with netaffx build 35\n",
      " - web_link : http://www.affymetrix.com/support/technical/byproduct.affx?product=hgu95, http://www.affymetrix.com/analysis/index.affx\n",
      " - contact_name : ,,Affymetrix, Inc.\n",
      " - contact_email : geo@ncbi.nlm.nih.gov, support@affymetrix.com\n",
      " - contact_phone : 888-362-2447\n",
      " - contact_institute : Affymetrix, Inc.\n",
      " - contact_address : \n",
      " - contact_city : Santa Clara\n",
      " - contact_state : CA\n",
      " - contact_zip/postal_code : 95051\n",
      " - contact_country : USA\n",
      " - contact_web_link : http://www.affymetrix.com/index.affx\n",
      " - data_row_count : 12625\n",
      "Table data:\n",
      "          ID  GB_ACC SPOT_ID Species Scientific Name Annotation Date  \\\n",
      "0    1000_at  X60188     NaN            Homo sapiens     Oct 6, 2014   \n",
      "1    1001_at  X60957     NaN            Homo sapiens     Oct 6, 2014   \n",
      "2  1002_f_at  X65962     NaN            Homo sapiens     Oct 6, 2014   \n",
      "3  1003_s_at  X68149     NaN            Homo sapiens     Oct 6, 2014   \n",
      "4    1004_at  X68149     NaN            Homo sapiens     Oct 6, 2014   \n",
      "\n",
      "       Sequence Type Sequence Source  \\\n",
      "0  Exemplar sequence         GenBank   \n",
      "1  Exemplar sequence         GenBank   \n",
      "2  Exemplar sequence         GenBank   \n",
      "3  Exemplar sequence         GenBank   \n",
      "4  Exemplar sequence         GenBank   \n",
      "\n",
      "                                  Target Description Representative Public ID  \\\n",
      "0  X60188 /FEATURE=mRNA /DEFINITION=HSERK1 Human ...                   X60188   \n",
      "1  X60957 /FEATURE=cds /DEFINITION=HSTIEMR Human ...                   X60957   \n",
      "2  X65962 /FEATURE=cds /DEFINITION=HSCP450 H.sapi...                   X65962   \n",
      "3  X68149 /FEATURE=cds /DEFINITION=HSBLR1A Homo s...                   X68149   \n",
      "4  X68149 /FEATURE=exon#2 /DEFINITION=HSBLR1A Hom...                   X68149   \n",
      "\n",
      "                                          Gene Title Gene Symbol  \\\n",
      "0                 mitogen-activated protein kinase 3       MAPK3   \n",
      "1  tyrosine kinase with immunoglobulin-like and E...        TIE1   \n",
      "2  cytochrome P450, family 2, subfamily C, polype...     CYP2C19   \n",
      "3                 chemokine (C-X-C motif) receptor 5       CXCR5   \n",
      "4                 chemokine (C-X-C motif) receptor 5       CXCR5   \n",
      "\n",
      "  ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
      "0           5595  NM_001040056 /// NM_001109891 /// NM_002746 //...   \n",
      "1           7075  NM_001253357 /// NM_005424 /// XM_005271163 //...   \n",
      "2           1557                                          NM_000769   \n",
      "3            643                            NM_001716 /// NM_032966   \n",
      "4            643                            NM_001716 /// NM_032966   \n",
      "\n",
      "                    Gene Ontology Biological Process  \\\n",
      "0  0000165 // MAPK cascade // non-traceable autho...   \n",
      "1  0001525 // angiogenesis // inferred from elect...   \n",
      "2  0006805 // xenobiotic metabolic process // tra...   \n",
      "3  0006928 // cellular component movement // trac...   \n",
      "4  0006928 // cellular component movement // trac...   \n",
      "\n",
      "                    Gene Ontology Cellular Component  \\\n",
      "0  0005634 // nucleus // inferred from direct ass...   \n",
      "1  0005886 // plasma membrane // inferred from el...   \n",
      "2  0005783 // endoplasmic reticulum // inferred f...   \n",
      "3  0005886 // plasma membrane // traceable author...   \n",
      "4  0005886 // plasma membrane // traceable author...   \n",
      "\n",
      "                    Gene Ontology Molecular Function  \n",
      "0  0000166 // nucleotide binding // inferred from...  \n",
      "1  0000166 // nucleotide binding // inferred from...  \n",
      "2  0004497 // monooxygenase activity // inferred ...  \n",
      "3  0004871 // signal transducer activity // infer...  \n",
      "4  0004871 // signal transducer activity // infer...  \n"
     ]
    }
   ],
   "source": [
    "# A GPL (or a Platform) contains a tab-delimited table containing the array definition eg. mappings from probe IDs to RefSeq IDs\n",
    "\n",
    "print()\n",
    "print(\"GPL example:\\n-------------\")\n",
    "for gpl_name, gpl in kidney_data.gpls.items():\n",
    "    print(\"Name: \", gpl_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gpl.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print(\"Table data:\",)\n",
    "    print(gpl.table.head())\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.a.\n",
    "Now your task is to load the data set from the study \"A circadian gene expression atlas in mammals assayed by microarray\" by Zhang et al (http://www.pnas.org/content/111/45/16219.long). The data is available in the GEO database (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE54650, accession ID 54650)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "28-Oct-2020 14:05:26 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "28-Oct-2020 14:05:26 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE54nnn/GSE54650/soft/GSE54650_family.soft.gz to ./GSE54650_family.soft.gz\n",
      "100%|█████████████████████████████████████████████████████████████████████████████| 65.9M/65.9M [00:46<00:00, 1.50MB/s]\n",
      "28-Oct-2020 14:06:14 DEBUG downloader - Size validation passed\n",
      "28-Oct-2020 14:06:14 DEBUG downloader - Moving C:\\Users\\Peder\\AppData\\Local\\Temp\\tmp5yyeo5ak to C:\\Users\\Peder\\Documents\\repos\\CBM-101\\CBM101\\C_Data_resources\\GSE54650_family.soft.gz\n",
      "28-Oct-2020 14:06:14 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE54nnn/GSE54650/soft/GSE54650_family.soft.gz\n",
      "28-Oct-2020 14:06:14 INFO GEOparse - Parsing ./GSE54650_family.soft.gz: \n",
      "28-Oct-2020 14:06:14 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "28-Oct-2020 14:06:14 DEBUG GEOparse - SERIES: GSE54650\n",
      "28-Oct-2020 14:06:14 DEBUG GEOparse - PLATFORM: GPL6246\n",
      "28-Oct-2020 14:06:19 DEBUG GEOparse - SAMPLE: GSM1320990\n",
      "28-Oct-2020 14:06:19 DEBUG GEOparse - SAMPLE: GSM1320991\n",
      "28-Oct-2020 14:06:19 DEBUG GEOparse - SAMPLE: GSM1320992\n",
      "28-Oct-2020 14:06:19 DEBUG GEOparse - SAMPLE: GSM1320993\n",
      "28-Oct-2020 14:06:20 DEBUG GEOparse - SAMPLE: GSM1320994\n",
      "28-Oct-2020 14:06:20 DEBUG GEOparse - SAMPLE: GSM1320995\n",
      "28-Oct-2020 14:06:20 DEBUG GEOparse - SAMPLE: GSM1320996\n",
      "28-Oct-2020 14:06:20 DEBUG GEOparse - SAMPLE: GSM1320997\n",
      "28-Oct-2020 14:06:20 DEBUG GEOparse - SAMPLE: GSM1320998\n",
      "28-Oct-2020 14:06:21 DEBUG GEOparse - SAMPLE: GSM1320999\n",
      "28-Oct-2020 14:06:21 DEBUG GEOparse - SAMPLE: GSM1321000\n",
      "28-Oct-2020 14:06:21 DEBUG GEOparse - SAMPLE: GSM1321001\n",
      "28-Oct-2020 14:06:21 DEBUG GEOparse - SAMPLE: GSM1321002\n",
      "28-Oct-2020 14:06:22 DEBUG GEOparse - SAMPLE: GSM1321003\n",
      "28-Oct-2020 14:06:22 DEBUG GEOparse - SAMPLE: GSM1321004\n",
      "28-Oct-2020 14:06:22 DEBUG GEOparse - SAMPLE: GSM1321005\n",
      "28-Oct-2020 14:06:22 DEBUG GEOparse - SAMPLE: GSM1321006\n",
      "28-Oct-2020 14:06:22 DEBUG GEOparse - SAMPLE: GSM1321007\n",
      "28-Oct-2020 14:06:23 DEBUG GEOparse - SAMPLE: GSM1321008\n",
      "28-Oct-2020 14:06:23 DEBUG GEOparse - SAMPLE: GSM1321009\n",
      "28-Oct-2020 14:06:23 DEBUG GEOparse - SAMPLE: GSM1321010\n",
      "28-Oct-2020 14:06:23 DEBUG GEOparse - SAMPLE: GSM1321011\n",
      "28-Oct-2020 14:06:23 DEBUG GEOparse - SAMPLE: GSM1321012\n",
      "28-Oct-2020 14:06:24 DEBUG GEOparse - SAMPLE: GSM1321013\n",
      "28-Oct-2020 14:06:24 DEBUG GEOparse - SAMPLE: GSM1321014\n",
      "28-Oct-2020 14:06:24 DEBUG GEOparse - SAMPLE: GSM1321015\n",
      "28-Oct-2020 14:06:24 DEBUG GEOparse - SAMPLE: GSM1321016\n",
      "28-Oct-2020 14:06:25 DEBUG GEOparse - SAMPLE: GSM1321017\n",
      "28-Oct-2020 14:06:25 DEBUG GEOparse - SAMPLE: GSM1321018\n",
      "28-Oct-2020 14:06:25 DEBUG GEOparse - SAMPLE: GSM1321019\n",
      "28-Oct-2020 14:06:25 DEBUG GEOparse - SAMPLE: GSM1321020\n",
      "28-Oct-2020 14:06:25 DEBUG GEOparse - SAMPLE: GSM1321021\n",
      "28-Oct-2020 14:06:26 DEBUG GEOparse - SAMPLE: GSM1321022\n",
      "28-Oct-2020 14:06:26 DEBUG GEOparse - SAMPLE: GSM1321023\n",
      "28-Oct-2020 14:06:26 DEBUG GEOparse - SAMPLE: GSM1321024\n",
      "28-Oct-2020 14:06:26 DEBUG GEOparse - SAMPLE: GSM1321025\n",
      "28-Oct-2020 14:06:26 DEBUG GEOparse - SAMPLE: GSM1321026\n",
      "28-Oct-2020 14:06:27 DEBUG GEOparse - SAMPLE: GSM1321027\n",
      "28-Oct-2020 14:06:27 DEBUG GEOparse - SAMPLE: GSM1321028\n",
      "28-Oct-2020 14:06:27 DEBUG GEOparse - SAMPLE: GSM1321029\n",
      "28-Oct-2020 14:06:27 DEBUG GEOparse - SAMPLE: GSM1321030\n",
      "28-Oct-2020 14:06:27 DEBUG GEOparse - SAMPLE: GSM1321031\n",
      "28-Oct-2020 14:06:28 DEBUG GEOparse - SAMPLE: GSM1321032\n",
      "28-Oct-2020 14:06:28 DEBUG GEOparse - SAMPLE: GSM1321033\n",
      "28-Oct-2020 14:06:28 DEBUG GEOparse - SAMPLE: GSM1321034\n",
      "28-Oct-2020 14:06:28 DEBUG GEOparse - SAMPLE: GSM1321035\n",
      "28-Oct-2020 14:06:29 DEBUG GEOparse - SAMPLE: GSM1321036\n",
      "28-Oct-2020 14:06:29 DEBUG GEOparse - SAMPLE: GSM1321037\n",
      "28-Oct-2020 14:06:29 DEBUG GEOparse - SAMPLE: GSM1321038\n",
      "28-Oct-2020 14:06:29 DEBUG GEOparse - SAMPLE: GSM1321039\n",
      "28-Oct-2020 14:06:29 DEBUG GEOparse - SAMPLE: GSM1321040\n",
      "28-Oct-2020 14:06:30 DEBUG GEOparse - SAMPLE: GSM1321041\n",
      "28-Oct-2020 14:06:30 DEBUG GEOparse - SAMPLE: GSM1321042\n",
      "28-Oct-2020 14:06:30 DEBUG GEOparse - SAMPLE: GSM1321043\n",
      "28-Oct-2020 14:06:30 DEBUG GEOparse - SAMPLE: GSM1321044\n",
      "28-Oct-2020 14:06:30 DEBUG GEOparse - SAMPLE: GSM1321045\n",
      "28-Oct-2020 14:06:31 DEBUG GEOparse - SAMPLE: GSM1321046\n",
      "28-Oct-2020 14:06:31 DEBUG GEOparse - SAMPLE: GSM1321047\n",
      "28-Oct-2020 14:06:31 DEBUG GEOparse - SAMPLE: GSM1321048\n",
      "28-Oct-2020 14:06:31 DEBUG GEOparse - SAMPLE: GSM1321049\n",
      "28-Oct-2020 14:06:32 DEBUG GEOparse - SAMPLE: GSM1321050\n",
      "28-Oct-2020 14:06:32 DEBUG GEOparse - SAMPLE: GSM1321051\n",
      "28-Oct-2020 14:06:32 DEBUG GEOparse - SAMPLE: GSM1321052\n",
      "28-Oct-2020 14:06:32 DEBUG GEOparse - SAMPLE: GSM1321053\n",
      "28-Oct-2020 14:06:32 DEBUG GEOparse - SAMPLE: GSM1321054\n",
      "28-Oct-2020 14:06:33 DEBUG GEOparse - SAMPLE: GSM1321055\n",
      "28-Oct-2020 14:06:33 DEBUG GEOparse - SAMPLE: GSM1321056\n",
      "28-Oct-2020 14:06:33 DEBUG GEOparse - SAMPLE: GSM1321057\n",
      "28-Oct-2020 14:06:34 DEBUG GEOparse - SAMPLE: GSM1321058\n",
      "28-Oct-2020 14:06:34 DEBUG GEOparse - SAMPLE: GSM1321059\n",
      "28-Oct-2020 14:06:34 DEBUG GEOparse - SAMPLE: GSM1321060\n",
      "28-Oct-2020 14:06:35 DEBUG GEOparse - SAMPLE: GSM1321061\n",
      "28-Oct-2020 14:06:35 DEBUG GEOparse - SAMPLE: GSM1321062\n",
      "28-Oct-2020 14:06:35 DEBUG GEOparse - SAMPLE: GSM1321063\n",
      "28-Oct-2020 14:06:35 DEBUG GEOparse - SAMPLE: GSM1321064\n",
      "28-Oct-2020 14:06:36 DEBUG GEOparse - SAMPLE: GSM1321065\n",
      "28-Oct-2020 14:06:36 DEBUG GEOparse - SAMPLE: GSM1321066\n",
      "28-Oct-2020 14:06:37 DEBUG GEOparse - SAMPLE: GSM1321067\n",
      "28-Oct-2020 14:06:37 DEBUG GEOparse - SAMPLE: GSM1321068\n",
      "28-Oct-2020 14:06:38 DEBUG GEOparse - SAMPLE: GSM1321069\n",
      "28-Oct-2020 14:06:38 DEBUG GEOparse - SAMPLE: GSM1321070\n",
      "28-Oct-2020 14:06:38 DEBUG GEOparse - SAMPLE: GSM1321071\n",
      "28-Oct-2020 14:06:39 DEBUG GEOparse - SAMPLE: GSM1321072\n",
      "28-Oct-2020 14:06:39 DEBUG GEOparse - SAMPLE: GSM1321073\n",
      "28-Oct-2020 14:06:39 DEBUG GEOparse - SAMPLE: GSM1321074\n",
      "28-Oct-2020 14:06:39 DEBUG GEOparse - SAMPLE: GSM1321075\n",
      "28-Oct-2020 14:06:40 DEBUG GEOparse - SAMPLE: GSM1321076\n",
      "28-Oct-2020 14:06:40 DEBUG GEOparse - SAMPLE: GSM1321077\n",
      "28-Oct-2020 14:06:40 DEBUG GEOparse - SAMPLE: GSM1321078\n",
      "28-Oct-2020 14:06:40 DEBUG GEOparse - SAMPLE: GSM1321079\n",
      "28-Oct-2020 14:06:40 DEBUG GEOparse - SAMPLE: GSM1321080\n",
      "28-Oct-2020 14:06:41 DEBUG GEOparse - SAMPLE: GSM1321081\n",
      "28-Oct-2020 14:06:41 DEBUG GEOparse - SAMPLE: GSM1321082\n",
      "28-Oct-2020 14:06:41 DEBUG GEOparse - SAMPLE: GSM1321083\n",
      "28-Oct-2020 14:06:41 DEBUG GEOparse - SAMPLE: GSM1321084\n",
      "28-Oct-2020 14:06:41 DEBUG GEOparse - SAMPLE: GSM1321085\n",
      "28-Oct-2020 14:06:42 DEBUG GEOparse - SAMPLE: GSM1321086\n",
      "28-Oct-2020 14:06:42 DEBUG GEOparse - SAMPLE: GSM1321087\n",
      "28-Oct-2020 14:06:42 DEBUG GEOparse - SAMPLE: GSM1321088\n",
      "28-Oct-2020 14:06:42 DEBUG GEOparse - SAMPLE: GSM1321089\n",
      "28-Oct-2020 14:06:42 DEBUG GEOparse - SAMPLE: GSM1321090\n",
      "28-Oct-2020 14:06:43 DEBUG GEOparse - SAMPLE: GSM1321091\n",
      "28-Oct-2020 14:06:43 DEBUG GEOparse - SAMPLE: GSM1321092\n",
      "28-Oct-2020 14:06:43 DEBUG GEOparse - SAMPLE: GSM1321093\n",
      "28-Oct-2020 14:06:43 DEBUG GEOparse - SAMPLE: GSM1321094\n",
      "28-Oct-2020 14:06:44 DEBUG GEOparse - SAMPLE: GSM1321095\n",
      "28-Oct-2020 14:06:44 DEBUG GEOparse - SAMPLE: GSM1321096\n",
      "28-Oct-2020 14:06:44 DEBUG GEOparse - SAMPLE: GSM1321097\n",
      "28-Oct-2020 14:06:44 DEBUG GEOparse - SAMPLE: GSM1321098\n",
      "28-Oct-2020 14:06:44 DEBUG GEOparse - SAMPLE: GSM1321099\n",
      "28-Oct-2020 14:06:45 DEBUG GEOparse - SAMPLE: GSM1321100\n",
      "28-Oct-2020 14:06:45 DEBUG GEOparse - SAMPLE: GSM1321101\n",
      "28-Oct-2020 14:06:45 DEBUG GEOparse - SAMPLE: GSM1321102\n",
      "28-Oct-2020 14:06:45 DEBUG GEOparse - SAMPLE: GSM1321103\n",
      "28-Oct-2020 14:06:46 DEBUG GEOparse - SAMPLE: GSM1321104\n",
      "28-Oct-2020 14:06:46 DEBUG GEOparse - SAMPLE: GSM1321105\n",
      "28-Oct-2020 14:06:46 DEBUG GEOparse - SAMPLE: GSM1321106\n",
      "28-Oct-2020 14:06:46 DEBUG GEOparse - SAMPLE: GSM1321107\n",
      "28-Oct-2020 14:06:46 DEBUG GEOparse - SAMPLE: GSM1321108\n",
      "28-Oct-2020 14:06:47 DEBUG GEOparse - SAMPLE: GSM1321109\n",
      "28-Oct-2020 14:06:47 DEBUG GEOparse - SAMPLE: GSM1321110\n",
      "28-Oct-2020 14:06:47 DEBUG GEOparse - SAMPLE: GSM1321111\n",
      "28-Oct-2020 14:06:47 DEBUG GEOparse - SAMPLE: GSM1321112\n",
      "28-Oct-2020 14:06:48 DEBUG GEOparse - SAMPLE: GSM1321113\n",
      "28-Oct-2020 14:06:48 DEBUG GEOparse - SAMPLE: GSM1321114\n",
      "28-Oct-2020 14:06:48 DEBUG GEOparse - SAMPLE: GSM1321115\n",
      "28-Oct-2020 14:06:48 DEBUG GEOparse - SAMPLE: GSM1321116\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "28-Oct-2020 14:06:48 DEBUG GEOparse - SAMPLE: GSM1321117\n",
      "28-Oct-2020 14:06:49 DEBUG GEOparse - SAMPLE: GSM1321118\n",
      "28-Oct-2020 14:06:49 DEBUG GEOparse - SAMPLE: GSM1321119\n",
      "28-Oct-2020 14:06:49 DEBUG GEOparse - SAMPLE: GSM1321120\n",
      "28-Oct-2020 14:06:49 DEBUG GEOparse - SAMPLE: GSM1321121\n",
      "28-Oct-2020 14:06:49 DEBUG GEOparse - SAMPLE: GSM1321122\n",
      "28-Oct-2020 14:06:50 DEBUG GEOparse - SAMPLE: GSM1321123\n",
      "28-Oct-2020 14:06:50 DEBUG GEOparse - SAMPLE: GSM1321124\n",
      "28-Oct-2020 14:06:50 DEBUG GEOparse - SAMPLE: GSM1321125\n",
      "28-Oct-2020 14:06:50 DEBUG GEOparse - SAMPLE: GSM1321126\n",
      "28-Oct-2020 14:06:50 DEBUG GEOparse - SAMPLE: GSM1321127\n",
      "28-Oct-2020 14:06:51 DEBUG GEOparse - SAMPLE: GSM1321128\n",
      "28-Oct-2020 14:06:51 DEBUG GEOparse - SAMPLE: GSM1321129\n",
      "28-Oct-2020 14:06:51 DEBUG GEOparse - SAMPLE: GSM1321130\n",
      "28-Oct-2020 14:06:51 DEBUG GEOparse - SAMPLE: GSM1321131\n",
      "28-Oct-2020 14:06:52 DEBUG GEOparse - SAMPLE: GSM1321132\n",
      "28-Oct-2020 14:06:52 DEBUG GEOparse - SAMPLE: GSM1321133\n",
      "28-Oct-2020 14:06:52 DEBUG GEOparse - SAMPLE: GSM1321134\n",
      "28-Oct-2020 14:06:52 DEBUG GEOparse - SAMPLE: GSM1321135\n",
      "28-Oct-2020 14:06:53 DEBUG GEOparse - SAMPLE: GSM1321136\n",
      "28-Oct-2020 14:06:53 DEBUG GEOparse - SAMPLE: GSM1321137\n",
      "28-Oct-2020 14:06:53 DEBUG GEOparse - SAMPLE: GSM1321138\n",
      "28-Oct-2020 14:06:53 DEBUG GEOparse - SAMPLE: GSM1321139\n",
      "28-Oct-2020 14:06:53 DEBUG GEOparse - SAMPLE: GSM1321140\n",
      "28-Oct-2020 14:06:54 DEBUG GEOparse - SAMPLE: GSM1321141\n",
      "28-Oct-2020 14:06:54 DEBUG GEOparse - SAMPLE: GSM1321142\n",
      "28-Oct-2020 14:06:54 DEBUG GEOparse - SAMPLE: GSM1321143\n",
      "28-Oct-2020 14:06:54 DEBUG GEOparse - SAMPLE: GSM1321144\n",
      "28-Oct-2020 14:06:55 DEBUG GEOparse - SAMPLE: GSM1321145\n",
      "28-Oct-2020 14:06:55 DEBUG GEOparse - SAMPLE: GSM1321146\n",
      "28-Oct-2020 14:06:55 DEBUG GEOparse - SAMPLE: GSM1321147\n",
      "28-Oct-2020 14:06:55 DEBUG GEOparse - SAMPLE: GSM1321148\n",
      "28-Oct-2020 14:06:55 DEBUG GEOparse - SAMPLE: GSM1321149\n",
      "28-Oct-2020 14:06:56 DEBUG GEOparse - SAMPLE: GSM1321150\n",
      "28-Oct-2020 14:06:56 DEBUG GEOparse - SAMPLE: GSM1321151\n",
      "28-Oct-2020 14:06:56 DEBUG GEOparse - SAMPLE: GSM1321152\n",
      "28-Oct-2020 14:06:56 DEBUG GEOparse - SAMPLE: GSM1321153\n",
      "28-Oct-2020 14:06:56 DEBUG GEOparse - SAMPLE: GSM1321154\n",
      "28-Oct-2020 14:06:57 DEBUG GEOparse - SAMPLE: GSM1321155\n",
      "28-Oct-2020 14:06:57 DEBUG GEOparse - SAMPLE: GSM1321156\n",
      "28-Oct-2020 14:06:57 DEBUG GEOparse - SAMPLE: GSM1321157\n",
      "28-Oct-2020 14:06:57 DEBUG GEOparse - SAMPLE: GSM1321158\n",
      "28-Oct-2020 14:06:57 DEBUG GEOparse - SAMPLE: GSM1321159\n",
      "28-Oct-2020 14:06:58 DEBUG GEOparse - SAMPLE: GSM1321160\n",
      "28-Oct-2020 14:06:58 DEBUG GEOparse - SAMPLE: GSM1321161\n",
      "28-Oct-2020 14:06:58 DEBUG GEOparse - SAMPLE: GSM1321162\n",
      "28-Oct-2020 14:06:58 DEBUG GEOparse - SAMPLE: GSM1321163\n",
      "28-Oct-2020 14:06:59 DEBUG GEOparse - SAMPLE: GSM1321164\n",
      "28-Oct-2020 14:06:59 DEBUG GEOparse - SAMPLE: GSM1321165\n",
      "28-Oct-2020 14:06:59 DEBUG GEOparse - SAMPLE: GSM1321166\n",
      "28-Oct-2020 14:06:59 DEBUG GEOparse - SAMPLE: GSM1321167\n",
      "28-Oct-2020 14:06:59 DEBUG GEOparse - SAMPLE: GSM1321168\n",
      "28-Oct-2020 14:07:00 DEBUG GEOparse - SAMPLE: GSM1321169\n",
      "28-Oct-2020 14:07:00 DEBUG GEOparse - SAMPLE: GSM1321170\n",
      "28-Oct-2020 14:07:00 DEBUG GEOparse - SAMPLE: GSM1321171\n",
      "28-Oct-2020 14:07:00 DEBUG GEOparse - SAMPLE: GSM1321172\n",
      "28-Oct-2020 14:07:00 DEBUG GEOparse - SAMPLE: GSM1321173\n",
      "28-Oct-2020 14:07:01 DEBUG GEOparse - SAMPLE: GSM1321174\n",
      "28-Oct-2020 14:07:01 DEBUG GEOparse - SAMPLE: GSM1321175\n",
      "28-Oct-2020 14:07:01 DEBUG GEOparse - SAMPLE: GSM1321176\n",
      "28-Oct-2020 14:07:01 DEBUG GEOparse - SAMPLE: GSM1321177\n",
      "28-Oct-2020 14:07:01 DEBUG GEOparse - SAMPLE: GSM1321178\n",
      "28-Oct-2020 14:07:02 DEBUG GEOparse - SAMPLE: GSM1321179\n",
      "28-Oct-2020 14:07:02 DEBUG GEOparse - SAMPLE: GSM1321180\n",
      "28-Oct-2020 14:07:02 DEBUG GEOparse - SAMPLE: GSM1321181\n",
      "28-Oct-2020 14:07:02 DEBUG GEOparse - SAMPLE: GSM1321182\n",
      "28-Oct-2020 14:07:02 DEBUG GEOparse - SAMPLE: GSM1321183\n",
      "28-Oct-2020 14:07:03 DEBUG GEOparse - SAMPLE: GSM1321184\n",
      "28-Oct-2020 14:07:03 DEBUG GEOparse - SAMPLE: GSM1321185\n",
      "28-Oct-2020 14:07:03 DEBUG GEOparse - SAMPLE: GSM1321186\n",
      "28-Oct-2020 14:07:03 DEBUG GEOparse - SAMPLE: GSM1321187\n",
      "28-Oct-2020 14:07:04 DEBUG GEOparse - SAMPLE: GSM1321188\n",
      "28-Oct-2020 14:07:04 DEBUG GEOparse - SAMPLE: GSM1321189\n",
      "28-Oct-2020 14:07:04 DEBUG GEOparse - SAMPLE: GSM1321190\n",
      "28-Oct-2020 14:07:04 DEBUG GEOparse - SAMPLE: GSM1321191\n",
      "28-Oct-2020 14:07:04 DEBUG GEOparse - SAMPLE: GSM1321192\n",
      "28-Oct-2020 14:07:05 DEBUG GEOparse - SAMPLE: GSM1321193\n",
      "28-Oct-2020 14:07:05 DEBUG GEOparse - SAMPLE: GSM1321194\n",
      "28-Oct-2020 14:07:05 DEBUG GEOparse - SAMPLE: GSM1321195\n",
      "28-Oct-2020 14:07:05 DEBUG GEOparse - SAMPLE: GSM1321196\n",
      "28-Oct-2020 14:07:05 DEBUG GEOparse - SAMPLE: GSM1321197\n",
      "28-Oct-2020 14:07:06 DEBUG GEOparse - SAMPLE: GSM1321198\n",
      "28-Oct-2020 14:07:06 DEBUG GEOparse - SAMPLE: GSM1321199\n",
      "28-Oct-2020 14:07:06 DEBUG GEOparse - SAMPLE: GSM1321200\n",
      "28-Oct-2020 14:07:06 DEBUG GEOparse - SAMPLE: GSM1321201\n",
      "28-Oct-2020 14:07:06 DEBUG GEOparse - SAMPLE: GSM1321202\n",
      "28-Oct-2020 14:07:07 DEBUG GEOparse - SAMPLE: GSM1321203\n",
      "28-Oct-2020 14:07:07 DEBUG GEOparse - SAMPLE: GSM1321204\n",
      "28-Oct-2020 14:07:07 DEBUG GEOparse - SAMPLE: GSM1321205\n",
      "28-Oct-2020 14:07:07 DEBUG GEOparse - SAMPLE: GSM1321206\n",
      "28-Oct-2020 14:07:08 DEBUG GEOparse - SAMPLE: GSM1321207\n",
      "28-Oct-2020 14:07:08 DEBUG GEOparse - SAMPLE: GSM1321208\n",
      "28-Oct-2020 14:07:08 DEBUG GEOparse - SAMPLE: GSM1321209\n",
      "28-Oct-2020 14:07:08 DEBUG GEOparse - SAMPLE: GSM1321210\n",
      "28-Oct-2020 14:07:09 DEBUG GEOparse - SAMPLE: GSM1321211\n",
      "28-Oct-2020 14:07:09 DEBUG GEOparse - SAMPLE: GSM1321212\n",
      "28-Oct-2020 14:07:09 DEBUG GEOparse - SAMPLE: GSM1321213\n",
      "28-Oct-2020 14:07:09 DEBUG GEOparse - SAMPLE: GSM1321214\n",
      "28-Oct-2020 14:07:10 DEBUG GEOparse - SAMPLE: GSM1321215\n",
      "28-Oct-2020 14:07:10 DEBUG GEOparse - SAMPLE: GSM1321216\n",
      "28-Oct-2020 14:07:10 DEBUG GEOparse - SAMPLE: GSM1321217\n",
      "28-Oct-2020 14:07:10 DEBUG GEOparse - SAMPLE: GSM1321218\n",
      "28-Oct-2020 14:07:10 DEBUG GEOparse - SAMPLE: GSM1321219\n",
      "28-Oct-2020 14:07:11 DEBUG GEOparse - SAMPLE: GSM1321220\n",
      "28-Oct-2020 14:07:11 DEBUG GEOparse - SAMPLE: GSM1321221\n",
      "28-Oct-2020 14:07:11 DEBUG GEOparse - SAMPLE: GSM1321222\n",
      "28-Oct-2020 14:07:11 DEBUG GEOparse - SAMPLE: GSM1321223\n",
      "28-Oct-2020 14:07:12 DEBUG GEOparse - SAMPLE: GSM1321224\n",
      "28-Oct-2020 14:07:12 DEBUG GEOparse - SAMPLE: GSM1321225\n",
      "28-Oct-2020 14:07:12 DEBUG GEOparse - SAMPLE: GSM1321226\n",
      "28-Oct-2020 14:07:12 DEBUG GEOparse - SAMPLE: GSM1321227\n",
      "28-Oct-2020 14:07:12 DEBUG GEOparse - SAMPLE: GSM1321228\n",
      "28-Oct-2020 14:07:13 DEBUG GEOparse - SAMPLE: GSM1321229\n",
      "28-Oct-2020 14:07:13 DEBUG GEOparse - SAMPLE: GSM1321230\n",
      "28-Oct-2020 14:07:13 DEBUG GEOparse - SAMPLE: GSM1321231\n",
      "28-Oct-2020 14:07:13 DEBUG GEOparse - SAMPLE: GSM1321232\n",
      "28-Oct-2020 14:07:13 DEBUG GEOparse - SAMPLE: GSM1321233\n",
      "28-Oct-2020 14:07:14 DEBUG GEOparse - SAMPLE: GSM1321234\n",
      "28-Oct-2020 14:07:14 DEBUG GEOparse - SAMPLE: GSM1321235\n",
      "28-Oct-2020 14:07:14 DEBUG GEOparse - SAMPLE: GSM1321236\n",
      "28-Oct-2020 14:07:14 DEBUG GEOparse - SAMPLE: GSM1321237\n",
      "28-Oct-2020 14:07:15 DEBUG GEOparse - SAMPLE: GSM1321238\n",
      "28-Oct-2020 14:07:15 DEBUG GEOparse - SAMPLE: GSM1321239\n",
      "28-Oct-2020 14:07:15 DEBUG GEOparse - SAMPLE: GSM1321240\n",
      "28-Oct-2020 14:07:15 DEBUG GEOparse - SAMPLE: GSM1321241\n",
      "28-Oct-2020 14:07:15 DEBUG GEOparse - SAMPLE: GSM1321242\n",
      "28-Oct-2020 14:07:16 DEBUG GEOparse - SAMPLE: GSM1321243\n",
      "28-Oct-2020 14:07:16 DEBUG GEOparse - SAMPLE: GSM1321244\n",
      "28-Oct-2020 14:07:16 DEBUG GEOparse - SAMPLE: GSM1321245\n",
      "28-Oct-2020 14:07:16 DEBUG GEOparse - SAMPLE: GSM1321246\n",
      "28-Oct-2020 14:07:17 DEBUG GEOparse - SAMPLE: GSM1321247\n",
      "28-Oct-2020 14:07:17 DEBUG GEOparse - SAMPLE: GSM1321248\n",
      "28-Oct-2020 14:07:17 DEBUG GEOparse - SAMPLE: GSM1321249\n",
      "28-Oct-2020 14:07:17 DEBUG GEOparse - SAMPLE: GSM1321250\n",
      "28-Oct-2020 14:07:17 DEBUG GEOparse - SAMPLE: GSM1321251\n",
      "28-Oct-2020 14:07:18 DEBUG GEOparse - SAMPLE: GSM1321252\n",
      "28-Oct-2020 14:07:18 DEBUG GEOparse - SAMPLE: GSM1321253\n",
      "28-Oct-2020 14:07:18 DEBUG GEOparse - SAMPLE: GSM1321254\n",
      "28-Oct-2020 14:07:18 DEBUG GEOparse - SAMPLE: GSM1321255\n",
      "28-Oct-2020 14:07:18 DEBUG GEOparse - SAMPLE: GSM1321256\n",
      "28-Oct-2020 14:07:19 DEBUG GEOparse - SAMPLE: GSM1321257\n",
      "28-Oct-2020 14:07:19 DEBUG GEOparse - SAMPLE: GSM1321258\n",
      "28-Oct-2020 14:07:19 DEBUG GEOparse - SAMPLE: GSM1321259\n",
      "28-Oct-2020 14:07:19 DEBUG GEOparse - SAMPLE: GSM1321260\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "28-Oct-2020 14:07:20 DEBUG GEOparse - SAMPLE: GSM1321261\n",
      "28-Oct-2020 14:07:20 DEBUG GEOparse - SAMPLE: GSM1321262\n",
      "28-Oct-2020 14:07:20 DEBUG GEOparse - SAMPLE: GSM1321263\n",
      "28-Oct-2020 14:07:20 DEBUG GEOparse - SAMPLE: GSM1321264\n",
      "28-Oct-2020 14:07:20 DEBUG GEOparse - SAMPLE: GSM1321265\n",
      "28-Oct-2020 14:07:21 DEBUG GEOparse - SAMPLE: GSM1321266\n",
      "28-Oct-2020 14:07:21 DEBUG GEOparse - SAMPLE: GSM1321267\n",
      "28-Oct-2020 14:07:21 DEBUG GEOparse - SAMPLE: GSM1321268\n",
      "28-Oct-2020 14:07:21 DEBUG GEOparse - SAMPLE: GSM1321269\n",
      "28-Oct-2020 14:07:22 DEBUG GEOparse - SAMPLE: GSM1321270\n",
      "28-Oct-2020 14:07:22 DEBUG GEOparse - SAMPLE: GSM1321271\n",
      "28-Oct-2020 14:07:22 DEBUG GEOparse - SAMPLE: GSM1321272\n",
      "28-Oct-2020 14:07:22 DEBUG GEOparse - SAMPLE: GSM1321273\n",
      "28-Oct-2020 14:07:22 DEBUG GEOparse - SAMPLE: GSM1321274\n",
      "28-Oct-2020 14:07:23 DEBUG GEOparse - SAMPLE: GSM1321275\n",
      "28-Oct-2020 14:07:23 DEBUG GEOparse - SAMPLE: GSM1321276\n",
      "28-Oct-2020 14:07:23 DEBUG GEOparse - SAMPLE: GSM1321277\n"
     ]
    }
   ],
   "source": [
    "# %load solutions/ex1_2a.py\n",
    "# download the data set using GEOparse:\n",
    "circadian_expression = GEOparse.get_GEO(geo=\"GSE54650\", destdir=\"./\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b) use the GSM example and GPL example codes above to print information of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GSM example:\n",
      "-------------\n",
      "Name:  GSM1320990\n",
      "Metadata:\n",
      " - title : Adr_CT18\n",
      " - geo_accession : GSM1320990\n",
      " - status : Public on Sep 30 2014\n",
      " - submission_date : Feb 04 2014\n",
      " - last_update_date : Sep 30 2014\n",
      " - type : RNA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : adrenal gland\n",
      " - organism_ch1 : Mus musculus\n",
      " - taxid_ch1 : 10090\n",
      " - characteristics_ch1 : strain: C57/BL6, tissue: adrenal gland\n",
      " - growth_protocol_ch1 : C57/BL6 mice were entrained to a 12h light : 12h dark schedule for one week. Mice were then transferred to complete darkness prior to collection. Collection began at CT18. Mice were housed in light-tight boxes for the duration of the experiment.\n",
      " - molecule_ch1 : total RNA\n",
      " - extract_protocol_ch1 : Organ samples were homogenized in Trizol reagent (Invitrogen) using a Tissuelyser (Qiagen). RNA was extracted using RNeasy columns (Qiagen) as per manufacturerâ€™s protocol, then pooled from three mice for each organ and time point.\n",
      " - label_ch1 : biotin\n",
      " - label_protocol_ch1 : not provided\n",
      " - hyb_protocol : not provided\n",
      " - scan_protocol : not provided\n",
      " - data_processing : Intensity values normalized using Affymetrix Expression Console software (GC-RMA).\n",
      " - platform_id : GPL6246\n",
      " - contact_name : Nicholas,,Lahens\n",
      " - contact_department : ITMAT\n",
      " - contact_institute : University of Pennsylvania\n",
      " - contact_address : Smilow Center for Translational Research 10-110 3400 Civic Center Blvd, Bldg 421\n",
      " - contact_city : Philadelphia\n",
      " - contact_state : PA\n",
      " - contact_zip/postal_code : 19104\n",
      " - contact_country : USA\n",
      " - supplementary_file : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1320nnn/GSM1320990/suppl/GSM1320990_Adr18_MoGene1.0ST.CEL.gz\n",
      " - series_id : GSE54650, GSE54652\n",
      " - data_row_count : 35556\n",
      "Table data:\n",
      "     ID_REF   VALUE\n",
      "0  10338001  2749.7\n",
      "1  10338002    64.2\n",
      "2  10338003   787.2\n",
      "3  10338004   431.0\n",
      "4  10338005    11.6\n",
      "\n",
      "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@\n",
      "\n",
      "GPL example:\n",
      "-------------\n",
      "Name:  GPL6246\n",
      "Metadata:\n",
      " - title : [MoGene-1_0-st] Affymetrix Mouse Gene 1.0 ST Array [transcript (gene) version]\n",
      " - geo_accession : GPL6246\n",
      " - status : Public on Dec 05 2007\n",
      " - submission_date : Dec 05 2007\n",
      " - last_update_date : Mar 04 2019\n",
      " - technology : in situ oligonucleotide\n",
      " - distribution : commercial\n",
      " - organism : Mus musculus\n",
      " - taxid : 10090\n",
      " - manufacturer : Affymetrix\n",
      " - manufacture_protocol : See manufacturer's web site, \n",
      " - description : Affymetrix submissions are typically submitted to GEO using the GEOarchive method described at http://www.ncbi.nlm.nih.gov/projects/geo/info/geo_affy.html, , June 03, 2009: annotation table updated with netaffx build 28, June 07, 2012: annotation table updated with netaffx build 32, July 01, 2016: annotation table updated with netaffx build 35\n",
      " - web_link : http://www.affymetrix.com/support/technical/byproduct.affx?product=mogene-1_0-st-v1, http://www.affymetrix.com/support/technical/libraryfilesmain.affx\n",
      " - contact_name : ,,Affymetrix, Inc.\n",
      " - contact_email : geo@ncbi.nlm.nih.gov, support@affymetrix.com\n",
      " - contact_phone : 888-362-2447\n",
      " - contact_institute : Affymetrix, Inc.\n",
      " - contact_address : \n",
      " - contact_city : Santa Clara\n",
      " - contact_state : CA\n",
      " - contact_zip/postal_code : 95051\n",
      " - contact_country : USA\n",
      " - contact_web_link : http://www.affymetrix.com/index.affx\n",
      " - relation : Alternative to: GPL9525 (Alternative CDF [affy9939alt]), Alternative to: GPL10528 (Alternative CDF  [MoGene10stv1_Mm_ENSG]), Alternative to: GPL10655 ( v1.na30.1.mm9.probeset), Alternative to: GPL10740 (probe set (exon) version), Alternative to: GPL10742 (Alternative CDF  [ MoGene10stv1_Mm_ENST]), Alternative to: GPL11078 (Alternative CDF [MoGene10stv1_Mm_ENTREZG_13]), Alternative to: GPL11204 (Alternative CDF [ MoGene10stv1_Mm_ENST]), Alternative to: GPL13455 (Alternative CDF [Brainarray MoGene10stv1_Mm_REFSEQ version 14]), Alternative to: GPL13730 (Alternative CDF [ MoGene10stv1_Mm_ENTREZG_14.1.0]), Alternative to: GPL14866 (Alternative CDF [MoGene10stv1_Mm_ENSG version 14]), Alternative to: GPL15733 (Alternative CDF [ MoGene10stv1_Mm_ENTREZG_12.1.0]), Alternative to: GPL15912 (Alternative CDF [ MoGene10stv1_Mm_REFSEQ_13.0.0]), Alternative to: GPL17103 (Alternative CDF [MoGene10stv1_Mm_ENSG version 15]), Alternative to: GPL17777 (Alternative CDF [mogene10st_Mm_ENTREZG_17.1.0]), Alternative to: GPL17940 (mogene10st_annotation.csv), Alternative to: GPL17984 (transcript (gene) version, ENSG 60), Alternative to: GPL18535 (Alternative CDF [MoGene10stv1_Mm_ENTREZG_15.1.0]), Alternative to: GPL19103 (Alternative CDF [mogene10st_Mm_ENTREZG_18.0.0]), Alternative to: GPL19192 (Alternative CDF [GATExplorer_Ensembl v57]), Alternative to: GPL19317 (Alternative CDF [ MoGene10stv1_Mm_ENSG v.18.0.0]), Alternative to: GPL19485 (Alternative CDF [MoGene10stv1_Mm_ENTREZG_16.0.0]), Alternative to: GPL19972 (Alternative CDF [MoGene10stv1_Mm_REFSEQ version 18]), Alternative to: GPL20023 (Alternative CDF [mogene10stv1_75_030genes]), Alternative to: GPL20205 ([MoGene10stv1_Mm_REFSEQ_14.1.0]), Alternative to: GPL20894 (Alternative CDF [MoGene10stv1_Mm_ENSG v.19.0.0]), Alternative to: GPL21783 (Alternative CDF[MoGene10st_Mm_ENTREZG_17.0.0]), Alternative to: GPL26217 (Alternative CDF [mogene10st_Mm_ENTREZG_22.0.0])\n",
      " - data_row_count : 35557\n",
      "Table data:\n",
      "         ID            GB_LIST               SPOT_ID seqname     RANGE_GB  \\\n",
      "0  10344614  AK145513,AK145782  chr1:3054233-3054733    chr1  NC_000067.6   \n",
      "1  10344616                NaN  chr1:3102016-3102125    chr1  NC_000067.6   \n",
      "2  10344618                NaN  chr1:3276323-3277348    chr1  NC_000067.6   \n",
      "3  10344620           AK140060  chr1:3680571-3680912    chr1  NC_000067.6   \n",
      "4  10344622                NaN  chr1:4771131-4772199    chr1  NC_000067.6   \n",
      "\n",
      "  RANGE_STRAND RANGE_START RANGE_STOP  total_probes  \\\n",
      "0            +     3054233    3054733          33.0   \n",
      "1            +     3102016    3102125          25.0   \n",
      "2            +     3276323    3277348          25.0   \n",
      "3            +     3680571    3680912          25.0   \n",
      "4            +     4771131    4772199           8.0   \n",
      "\n",
      "                                     gene_assignment  \\\n",
      "0  ENSMUST00000160944 // Gm16088 // predicted gen...   \n",
      "1  ENSMUST00000082908 // Gm26206 // predicted gen...   \n",
      "2                                                ---   \n",
      "3  AK140060 // Gm10568 // predicted gene 10568 //...   \n",
      "4                                                ---   \n",
      "\n",
      "                                     mrna_assignment category  \n",
      "0  ENSMUST00000160944 // ENSEMBL // havana:known ...     main  \n",
      "1  ENSMUST00000082908 // ENSEMBL // ncrna:known c...     main  \n",
      "2                                                ---     main  \n",
      "3  AK140060 // GenBank HTC // Mus musculus adult ...     main  \n",
      "4  GENSCAN00000005057 // ENSEMBL // cdna:genscan ...     main  \n"
     ]
    }
   ],
   "source": [
    "# %load solutions/ex1_2b.py\n",
    "\n",
    "# we only have to change the name of the data variable\n",
    "\n",
    "print()\n",
    "print(\"GSM example:\\n-------------\")\n",
    "for gsm_name, gsm in circadian_expression.gsms.items():\n",
    "    print(\"Name: \", gsm_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gsm.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print (\"Table data:\",)\n",
    "    print (gsm.table.head())\n",
    "    break\n",
    "\n",
    "    \n",
    "print('\\n@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
    "    \n",
    "    \n",
    "# GLP example:\n",
    "print()\n",
    "print(\"GPL example:\\n-------------\")\n",
    "for gpl_name, gpl in circadian_expression.gpls.items():\n",
    "    print(\"Name: \", gpl_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gpl.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print(\"Table data:\",)\n",
    "    print(gpl.table.head())\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CBM101",
   "language": "python",
   "name": "cbm101"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
